### ISSUE TITLE
ROCm 3.5.0 test: TestCArray32BitBoundary

### ISSUE THREAD

**Issue by leofang** [cupy/cupy#3833]
When working on fixing the tests for ROCm 3.5.0 (ex: #3823), I noticed one weird error:
```
$ pytest tests/cupy_tests/core_tests/test_carray.py 
========================================================================= test session starts =========================================================================
platform linux -- Python 3.7.8, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/leo/dev/cupy_rocm35, configfile: setup.cfg
collected 11 items                                                                                                                                                    

tests/cupy_tests/core_tests/test_carray.py ....F......                                                                                                          [100%]

============================================================================== FAILURES ===============================================================================
_______________________________________________________ TestCArray32BitBoundary_param_4_{size=4294967296}.test ________________________________________________________

self = <cupy.testing._bundle.TestCArray32BitBoundary_param_4_{size=4294967296} testMethod=test>

    def test(self):
        # Elementwise
        a = cupy.full((1, self.size), 7, dtype=cupy.int8)
        # Reduction
        result = a.sum(axis=0, dtype=cupy.int8)
>       assert result.sum() == self.size * 7
E       AssertionError: assert array(-283467841536) == (4294967296 * 7)
E        +  where array(-283467841536) = <built-in method sum of cupy.core.core.ndarray object at 0x7fdaf0c12e10>()
E        +    where <built-in method sum of cupy.core.core.ndarray object at 0x7fdaf0c12e10> = array([-66, -66, -66, ..., -66, -66, -66], dtype=int8).sum
E        +  and   4294967296 = <cupy.testing._bundle.TestCArray32BitBoundary_param_4_{size=4294967296} testMethod=test>.size

tests/cupy_tests/core_tests/test_carray.py:75: AssertionError
======================================================================= short test summary info =======================================================================
FAILED tests/cupy_tests/core_tests/test_carray.py::TestCArray32BitBoundary_param_4_{size=4294967296}::test - AssertionError: assert array(-283467841536) == (4294967...
==================================================================== 1 failed, 10 passed in 38.88s ====================================================================
```
I don't understand why it would fail when the array size is precisely 2**32 but fine with larger or smaller sizes than that.

Not sure if this is relevant, but this bool test fails if array size > 2**31:

<details><summary>(click here for a long output)</summary>
<p>

```python
>>> import cupy
>>> a = cupy.full((1, 2**32), 7, dtype=cupy.int8)
>>> a = a.sum(axis=0, dtype=cupy.int8)
>>> a[0]==7
HIPCC compilation error: `hipcc` command returns non-zero exit status. 
command: ['hipcc', '--genco', '-I/home/leo/dev/cupy_rocm35/cupy/core/include', '/tmp/tmp9lzrbqfr/kern.cpp', '-o', '/tmp/tmp9lzrbqfr/kern.hsaco']
return-code: 1
stdout/stderr: 
/tmp/tmp9lzrbqfr/kern.cpp:15:36: error: no viable overloaded operator[] for type 'const CArray<signed char, 0, 1, 0>'
        const in0_type in0(_raw_in0[_ind.get()]);
                           ~~~~~~~~^~~~~~~~~~~
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:202:17: note: candidate template ignored: could not match 'const Int [0]' against 'const ptrdiff_t *' (aka 'const long *')
  __device__ T& operator[](const Int (&idx)[ndim]) {
                ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:207:23: note: candidate template ignored: could not match 'const Int [0]' against 'const ptrdiff_t *' (aka 'const long *')
  __device__ const T& operator[](const Int (&idx)[ndim]) const {
                      ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:222:17: note: candidate function not viable: 'this' argument has type 'const CArray<signed char, 0, 1, 0>', but method is not marked const
  __device__ T& operator[](ptrdiff_t i) {
                ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:226:23: note: candidate function not viable: no known conversion from 'const ptrdiff_t *' (aka 'const long *') to 'ptrdiff_t' (aka 'long') for 1st argument; dereference the argument with *
  __device__ const T& operator[](ptrdiff_t idx) const {
                      ^
1 error generated when compiling for gfx906.

-----
Name: /tmp/tmp9lzrbqfr/kern.cpp
Options: -I/home/leo/dev/cupy_rocm35/cupy/core/include
CUDA source:
01 #include <hip/hip_runtime.h>
02 #include <cupy/complex.cuh>
03 #include <cupy/carray.cuh>
04 #include <cupy/atomics.cuh>
05 
06     typedef signed char in0_type;
07 typedef signed char in1_type;
08 typedef bool out0_type;
09 
10 
11     extern "C" __global__ void cupy_equal__int8_int_bool(const CArray<signed char, 0, 1, 0> _raw_in0, const signed char in1, CArray<bool, 0, 1, 1> _raw_out0, CIndexer<0> _ind) {
12       ;
13       CUPY_FOR(i, _ind.size()) {
14         _ind.set(i);
15         const in0_type in0(_raw_in0[_ind.get()]);
16 out0_type &out0 = _raw_out0[_ind.get()];
17 out0 = in0 == in1;
18       }
19       ;
20     }
21 
-----
Traceback (most recent call last):
  File "/home/leo/dev/cupy_rocm35/cupy/cuda/compiler.py", line 37, in _run_cc
    universal_newlines=True)
  File "/home/leo/miniconda3/envs/cupy_rocm35_dev/lib/python3.7/subprocess.py", line 411, in check_output
    **kwargs).stdout
  File "/home/leo/miniconda3/envs/cupy_rocm35_dev/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['hipcc', '--genco', '-I/home/leo/dev/cupy_rocm35/cupy/core/include', '/tmp/tmp9lzrbqfr/kern.cpp', '-o', '/tmp/tmp9lzrbqfr/kern.hsaco']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/leo/dev/cupy_rocm35/cupy/cuda/compiler.py", line 564, in compile_using_hipcc
    output = _run_cc(cmd, root_dir, 'hipcc', log_stream)
  File "/home/leo/dev/cupy_rocm35/cupy/cuda/compiler.py", line 53, in _run_cc
    raise HIPCCException(msg)
cupy.cuda.compiler.HIPCCException: `hipcc` command returns non-zero exit status. 
command: ['hipcc', '--genco', '-I/home/leo/dev/cupy_rocm35/cupy/core/include', '/tmp/tmp9lzrbqfr/kern.cpp', '-o', '/tmp/tmp9lzrbqfr/kern.hsaco']
return-code: 1
stdout/stderr: 
/tmp/tmp9lzrbqfr/kern.cpp:15:36: error: no viable overloaded operator[] for type 'const CArray<signed char, 0, 1, 0>'
        const in0_type in0(_raw_in0[_ind.get()]);
                           ~~~~~~~~^~~~~~~~~~~
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:202:17: note: candidate template ignored: could not match 'const Int [0]' against 'const ptrdiff_t *' (aka 'const long *')
  __device__ T& operator[](const Int (&idx)[ndim]) {
                ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:207:23: note: candidate template ignored: could not match 'const Int [0]' against 'const ptrdiff_t *' (aka 'const long *')
  __device__ const T& operator[](const Int (&idx)[ndim]) const {
                      ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:222:17: note: candidate function not viable: 'this' argument has type 'const CArray<signed char, 0, 1, 0>', but method is not marked const
  __device__ T& operator[](ptrdiff_t i) {
                ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:226:23: note: candidate function not viable: no known conversion from 'const ptrdiff_t *' (aka 'const long *') to 'ptrdiff_t' (aka 'long') for 1st argument; dereference the argument with *
  __device__ const T& operator[](ptrdiff_t idx) const {
                      ^
1 error generated when compiling for gfx906.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cupy/core/core.pyx", line 968, in cupy.core.core.ndarray.__richcmp__
    return _logic._ndarray_equal(self, other)
  File "cupy/core/_routines_logic.pyx", line 32, in cupy.core._routines_logic._ndarray_equal
    return _equal(self, other)
  File "cupy/core/_kernel.pyx", line 1084, in cupy.core._kernel.ufunc.__call__
    kern = self._get_ufunc_kernel(dev_id, op, arginfos)
  File "cupy/core/_kernel.pyx", line 1107, in cupy.core._kernel.ufunc._get_ufunc_kernel
    kern = _get_ufunc_kernel(
  File "cupy/core/_kernel.pyx", line 881, in cupy.core._kernel._get_ufunc_kernel
    return _get_simple_elementwise_kernel(
  File "cupy/core/_kernel.pyx", line 60, in cupy.core._kernel._get_simple_elementwise_kernel
    module = compile_with_cache(module_code, options)
  File "cupy/core/core.pyx", line 1875, in cupy.core.core.compile_with_cache
    return cuda.compile_with_cache(
  File "/home/leo/dev/cupy_rocm35/cupy/cuda/compiler.py", line 332, in compile_with_cache
    log_stream)
  File "/home/leo/dev/cupy_rocm35/cupy/cuda/compiler.py", line 662, in _compile_with_cache_hip
    binary = compile_using_hipcc(source, options, arch, log_stream)
  File "/home/leo/dev/cupy_rocm35/cupy/cuda/compiler.py", line 574, in compile_using_hipcc
    raise cex
cupy.cuda.compiler.CompileException: `hipcc` command returns non-zero exit status. 
command: ['hipcc', '--genco', '-I/home/leo/dev/cupy_rocm35/cupy/core/include', '/tmp/tmp9lzrbqfr/kern.cpp', '-o', '/tmp/tmp9lzrbqfr/kern.hsaco']
return-code: 1
stdout/stderr: 
/tmp/tmp9lzrbqfr/kern.cpp:15:36: error: no viable overloaded operator[] for type 'const CArray<signed char, 0, 1, 0>'
        const in0_type in0(_raw_in0[_ind.get()]);
                           ~~~~~~~~^~~~~~~~~~~
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:202:17: note: candidate template ignored: could not match 'const Int [0]' against 'const ptrdiff_t *' (aka 'const long *')
  __device__ T& operator[](const Int (&idx)[ndim]) {
                ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:207:23: note: candidate template ignored: could not match 'const Int [0]' against 'const ptrdiff_t *' (aka 'const long *')
  __device__ const T& operator[](const Int (&idx)[ndim]) const {
                      ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:222:17: note: candidate function not viable: 'this' argument has type 'const CArray<signed char, 0, 1, 0>', but method is not marked const
  __device__ T& operator[](ptrdiff_t i) {
                ^
/home/leo/dev/cupy_rocm35/cupy/core/include/cupy/carray.cuh:226:23: note: candidate function not viable: no known conversion from 'const ptrdiff_t *' (aka 'const long *') to 'ptrdiff_t' (aka 'long') for 1st argument; dereference the argument with *
  __device__ const T& operator[](ptrdiff_t idx) const {
                      ^
1 error generated when compiling for gfx906.
```

</p>
</details>

Wild guess: `ptrdiff_t` has different bounds on ROCm?!

**Comment by leofang** [comment]
Tried #3859 and it didn't help.

**Comment by leofang** [comment]
I checked that #4461 is the root cause. In the failing test, the grid size is 2**32/256 = 16777216 and the block size is 256.

**Comment by emcastillo** [comment]
Fixed by #4461

**Comment by leofang** [comment]
Ah, thanks! Sorry I missed it...
